File: controllers/appAuth.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const { AppUser, AppUserToken, HouseholdMember, Household } = require('../models');
const { Op } = require('sequelize');

class AppAuthController {
  /**
   * Sign in app user
   */
  async signIn(ctx) {
    try {
      const { username, password } = ctx.request.body;

      // Validate input
      if (!username || !password) {
        global.log.warn('APP_LOGIN_FAILED', 'Invalid app login attempt', {
          username
        }, {
          ipAddress: ctx.ip,
          userAgent: ctx.headers['user-agent']
        });

        ctx.status = 400;
        ctx.body = { error: 'Username and password are required' };
        return;
      }

      // Find app user by username
      const appUser = await AppUser.findOne({ where: { username } });

      if (!appUser) {
        global.log.warn('APP_LOGIN_FAILED', 'Invalid app login attempt', {
          username
        }, {
          ipAddress: ctx.ip,
          userAgent: ctx.headers['user-agent']
        });

        ctx.status = 401;
        ctx.body = { error: 'Invalid credentials' };
        return;
      }

      // Check if user is active
      if (!appUser.isActive) {
        global.log.warn('APP_LOGIN_FAILED', 'Invalid app login attempt', {
          username
        }, {
          ipAddress: ctx.ip,
          userAgent: ctx.headers['user-agent']
        });

        ctx.status = 401;
        ctx.body = { error: 'Account is deactivated' };
        return;
      }

      // Verify password
      const isValidPassword = await bcrypt.compare(password, appUser.password);

      if (!isValidPassword) {
        global.log.warn('APP_LOGIN_FAILED', 'Invalid app login attempt', {
          username
        }, {
          ipAddress: ctx.ip,
          userAgent: ctx.headers['user-agent']
        });

        ctx.status = 401;
        ctx.body = { error: 'Invalid credentials' };
        return;
      }

      // Update last login
      await appUser.update({ lastLogin: new Date() });

      // Generate access token
      const accessToken = jwt.sign(
        { id: appUser.id, username: appUser.username, type: 'appUser' },
        global.cfg.jwt.secret,
        { expiresIn: '1h' }
      );

      // Generate refresh token
      const refreshToken = crypto.randomBytes(64).toString('hex');

      // Hash refresh token before storing
      const hashedRefreshToken = await bcrypt.hash(refreshToken, global.cfg.bcrypt.saltRounds);

      // Store hashed refresh token with 7-day expiry
      const expiresAt = new Date();
      expiresAt.setDate(expiresAt.getDate() + 7);

      await AppUserToken.create({
        appUserId: appUser.id,
        token: hashedRefreshToken,
        expiresAt
      });

      global.log.info('APP_USER_LOGIN', 'App user logged in', {
        username: appUser.username
      }, {
        userId: appUser.id,
        ipAddress: ctx.ip,
        userAgent: ctx.headers['user-agent']
      });

      const memberships = await HouseholdMember.findAll({
        where: { appUserId: appUser.id },
        include: [{ model: Household, attributes: ['id', 'name'] }]
      });

      const households = memberships.map(m => ({
        householdId: m.householdId,
        householdName: m.Household?.name ?? '',
        role: m.role
      }));

      ctx.body = {
        success: true,
        accessToken,
        refreshToken,
        user: { ...appUser.toJSON(), households },
        expiresIn: 3600
      };
    } catch (error) {
      console.error('App sign in error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to sign in' };
    }
  }

  /**
   * Refresh access token using refresh token
   */
  async refreshToken(ctx) {
    try {
      const { refreshToken, userId } = ctx.request.body;

      // Validate input
      if (!refreshToken || !userId) {
        ctx.status = 400;
        ctx.body = { error: 'Refresh token and userId are required' };
        return;
      }

      // Find all non-expired tokens for this user
      const tokens = await AppUserToken.findAll({
        where: {
          appUserId: userId,
          expiresAt: { [Op.gt]: new Date() }
        }
      });

      if (!tokens || tokens.length === 0) {
        ctx.status = 401;
        ctx.body = { error: 'Invalid or expired refresh token' };
        return;
      }

      // Compare raw refresh token against each hashed one
      let matchedToken = null;
      for (const tokenRecord of tokens) {
        const isMatch = await bcrypt.compare(refreshToken, tokenRecord.token);
        if (isMatch) {
          matchedToken = tokenRecord;
          break;
        }
      }

      if (!matchedToken) {
        ctx.status = 401;
        ctx.body = { error: 'Invalid or expired refresh token' };
        return;
      }

      // Load the app user
      const appUser = await AppUser.findByPk(userId);

      if (!appUser || !appUser.isActive) {
        ctx.status = 401;
        ctx.body = { error: 'User not found or deactivated' };
        return;
      }

      // Delete the old token (rotate)
      await matchedToken.destroy();

      // Generate new access token
      const accessToken = jwt.sign(
        { id: appUser.id, username: appUser.username, type: 'appUser' },
        global.cfg.jwt.secret,
        { expiresIn: '1h' }
      );

      // Generate new refresh token
      const newRefreshToken = crypto.randomBytes(64).toString('hex');

      // Hash new refresh token before storing
      const hashedRefreshToken = await bcrypt.hash(newRefreshToken, global.cfg.bcrypt.saltRounds);

      // Store new hashed refresh token with 7-day expiry
      const expiresAt = new Date();
      expiresAt.setDate(expiresAt.getDate() + 7);

      await AppUserToken.create({
        appUserId: appUser.id,
        token: hashedRefreshToken,
        expiresAt
      });

      // Update last login
      await appUser.update({ lastLogin: new Date() });

      ctx.body = {
        success: true,
        accessToken,
        refreshToken: newRefreshToken,
        expiresIn: 3600
      };
    } catch (error) {
      console.error('App refresh token error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to refresh token' };
    }
  }

  /**
   * Get app user profile with households
   */
  async getProfile(ctx) {
    try {
      const appUser = ctx.state.appUser;

      // Load household memberships with household details
      const memberships = await HouseholdMember.findAll({
        where: { appUserId: appUser.id },
        include: [
          {
            model: Household,
            attributes: ['id', 'name']
          }
        ]
      });

      const households = memberships.map(m => ({
        householdId: m.Household.id,
        householdName: m.Household.name,
        role: m.role
      }));

      ctx.body = {
        success: true,
        user: appUser.toJSON(),
        households
      };
    } catch (error) {
      console.error('App get profile error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to get profile' };
    }
  }

  /**
   * Sign out app user by deleting refresh token
   */
  async signOut(ctx) {
    try {
      const { refreshToken, userId } = ctx.request.body;

      // Validate input
      if (!refreshToken || !userId) {
        ctx.status = 400;
        ctx.body = { error: 'Refresh token and userId are required' };
        return;
      }

      // Find all non-expired tokens for this user
      const tokens = await AppUserToken.findAll({
        where: { appUserId: userId }
      });

      // Find and delete the matching token
      let deleted = false;
      for (const tokenRecord of tokens) {
        const isMatch = await bcrypt.compare(refreshToken, tokenRecord.token);
        if (isMatch) {
          await tokenRecord.destroy();
          deleted = true;
          break;
        }
      }

      global.log.info('APP_USER_LOGOUT', 'App user signed out', {}, {
        userId,
        ipAddress: ctx.ip,
        userAgent: ctx.headers['user-agent']
      });

      ctx.body = { success: true };
    } catch (error) {
      console.error('App sign out error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to sign out' };
    }
  }
}

module.exports = new AppAuthController();

File: backend_go/internal/handlers/app_auth.go
package handlers

import (
	"crypto/rand"
	"encoding/hex"
	"household-go/internal/config"
	"household-go/internal/database"
	"household-go/internal/models"
	"household-go/internal/utils"
	"log"
	"time"

	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

// AppAuthHandler handles sign-in, token refresh, sign-out, and profile for AppUsers.
type AppAuthHandler struct {
	cfg *config.Config
}

func NewAppAuthHandler(cfg *config.Config) *AppAuthHandler {
	return &AppAuthHandler{cfg: cfg}
}

// SignIn authenticates an AppUser, returns access + refresh tokens, and household memberships.
// POST /api/app/auth/signin
func (h *AppAuthHandler) SignIn(c *gin.Context) {
	var body struct {
		Username string `json:"username" binding:"required"`
		Password string `json:"password" binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Username and password are required")
		return
	}

	var appUser models.AppUser
	if err := database.DB.Where("username = ?", body.Username).First(&appUser).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.Unauthorized(c, "Invalid credentials")
		} else {
			utils.InternalError(c, "Failed to sign in")
		}
		return
	}

	if !appUser.IsActive {
		utils.Unauthorized(c, "Account is deactivated")
		return
	}

	if err := bcrypt.CompareHashAndPassword([]byte(appUser.Password), []byte(body.Password)); err != nil {
		utils.Unauthorized(c, "Invalid credentials")
		return
	}

	// Update last login timestamp
	database.DB.Model(&appUser).Update("last_login", time.Now())

	// Generate short-lived access token (1 hour)
	accessToken, err := utils.GenerateAppToken(appUser.ID, appUser.Username, h.cfg.JWT.Secret)
	if err != nil {
		utils.InternalError(c, "Failed to sign in")
		return
	}

	// Generate refresh token: 64 random bytes encoded as hex = 128-char string.
	// crypto/rand is cryptographically secure (unlike math/rand which is NOT).
	// This is the Go equivalent of Node's `crypto.randomBytes(64).toString('hex')`.
	rawToken, err := generateSecureToken(64)
	if err != nil {
		utils.InternalError(c, "Failed to sign in")
		return
	}

	// Hash the refresh token before storing — never store raw tokens in the DB.
	// If the DB is leaked, hashed tokens are useless to an attacker.
	hashedToken, err := bcrypt.GenerateFromPassword([]byte(rawToken), h.cfg.Bcrypt.SaltRounds)
	if err != nil {
		utils.InternalError(c, "Failed to sign in")
		return
	}

	expiresAt := time.Now().AddDate(0, 0, 7) // 7 days
	tokenRecord := models.AppUserToken{
		AppUserID: appUser.ID,
		Token:     string(hashedToken),
		ExpiresAt: expiresAt,
	}
	if err := database.DB.Create(&tokenRecord).Error; err != nil {
		utils.InternalError(c, "Failed to sign in")
		return
	}

	// Load household memberships — equivalent to the Sequelize include in appAuth.js
	var memberships []models.HouseholdMember
	database.DB.Where("app_user_id = ?", appUser.ID).
		Preload("Household").
		Find(&memberships)

	// Build the households slice for the response
	type householdSummary struct {
		HouseholdID   uint   `json:"householdId"`
		HouseholdName string `json:"householdName"`
		Role          string `json:"role"`
	}
	households := make([]householdSummary, 0, len(memberships))
	for _, m := range memberships {
		households = append(households, householdSummary{
			HouseholdID:   m.HouseholdID,
			HouseholdName: m.Household.Name,
			Role:          m.Role,
		})
	}

	utils.OK(c, gin.H{
		"success":      true,
		"accessToken":  accessToken,
		"refreshToken": rawToken, // Send the RAW token to client; we stored the hash
		"expiresIn":    3600,
		"user":         appUser,
		"households":   households,
	})
}

// RefreshToken rotates the refresh token: validates old one, issues new access + refresh pair.
// POST /api/app/auth/refresh
func (h *AppAuthHandler) RefreshToken(c *gin.Context) {
	var body struct {
		RefreshToken string `json:"refreshToken" binding:"required"`
		UserID       uint   `json:"userId"       binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Refresh token and userId are required")
		return
	}

	// Load all non-expired tokens for this user, then bcrypt-compare each one.
	// We can't query by the raw token because we only stored the hash.
	// This is O(n) in the number of active sessions — acceptable for household scale.
	var tokens []models.AppUserToken
	database.DB.Where("app_user_id = ? AND expires_at > ?", body.UserID, time.Now()).Find(&tokens)

	if len(tokens) == 0 {
		utils.Unauthorized(c, "Invalid or expired refresh token")
		return
	}

	var matched *models.AppUserToken
	for i := range tokens {
		if err := bcrypt.CompareHashAndPassword([]byte(tokens[i].Token), []byte(body.RefreshToken)); err == nil {
			matched = &tokens[i]
			break
		}
	}

	if matched == nil {
		utils.Unauthorized(c, "Invalid or expired refresh token")
		return
	}

	var appUser models.AppUser
	if err := database.DB.First(&appUser, body.UserID).Error; err != nil || !appUser.IsActive {
		utils.Unauthorized(c, "User not found or deactivated")
		return
	}

	// Rotate: delete the old token record (prevents replay attacks)
	database.DB.Delete(matched)

	// Issue new access token
	accessToken, err := utils.GenerateAppToken(appUser.ID, appUser.Username, h.cfg.JWT.Secret)
	if err != nil {
		utils.InternalError(c, "Failed to refresh token")
		return
	}

	// Issue new refresh token
	rawToken, err := generateSecureToken(64)
	if err != nil {
		utils.InternalError(c, "Failed to refresh token")
		return
	}
	hashedToken, err := bcrypt.GenerateFromPassword([]byte(rawToken), h.cfg.Bcrypt.SaltRounds)
	if err != nil {
		utils.InternalError(c, "Failed to refresh token")
		return
	}

	database.DB.Create(&models.AppUserToken{
		AppUserID: appUser.ID,
		Token:     string(hashedToken),
		ExpiresAt: time.Now().AddDate(0, 0, 7),
	})

	database.DB.Model(&appUser).Update("last_login", time.Now())

	utils.OK(c, gin.H{
		"success":      true,
		"accessToken":  accessToken,
		"refreshToken": rawToken,
		"expiresIn":    3600,
	})
}

// SignOut invalidates the refresh token for the AppUser.
// POST /api/app/auth/signout
func (h *AppAuthHandler) SignOut(c *gin.Context) {
	var body struct {
		RefreshToken string `json:"refreshToken" binding:"required"`
		UserID       uint   `json:"userId"       binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Refresh token and userId are required")
		return
	}

	var tokens []models.AppUserToken
	database.DB.Where("app_user_id = ?", body.UserID).Find(&tokens)

	for i := range tokens {
		if err := bcrypt.CompareHashAndPassword([]byte(tokens[i].Token), []byte(body.RefreshToken)); err == nil {
			database.DB.Delete(&tokens[i])
			break
		}
	}

	utils.OK(c, gin.H{"success": true})
}

// GetProfile returns the AppUser's profile and their household memberships.
// GET /api/app/auth/profile
func (h *AppAuthHandler) GetProfile(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	var memberships []models.HouseholdMember
	database.DB.Where("app_user_id = ?", appUser.ID).Preload("Household").Find(&memberships)

	type householdSummary struct {
		HouseholdID   uint   `json:"householdId"`
		HouseholdName string `json:"householdName"`
		Role          string `json:"role"`
	}
	households := make([]householdSummary, 0, len(memberships))
	for _, m := range memberships {
		households = append(households, householdSummary{
			HouseholdID:   m.HouseholdID,
			HouseholdName: m.Household.Name,
			Role:          m.Role,
		})
	}

	utils.OK(c, gin.H{"success": true, "user": appUser, "households": households})
}

// generateSecureToken creates a cryptographically random hex string of byteLen*2 chars.
// crypto/rand.Read fills the buffer with random bytes from the OS entropy source.
func generateSecureToken(byteLen int) (string, error) {
	buf := make([]byte, byteLen)
	if _, err := rand.Read(buf); err != nil {
		log.Printf("crypto/rand.Read failed: %v", err)
		return "", err
	}
	return hex.EncodeToString(buf), nil
}
